### ISSUE TITLE
port forwarding doesn't work on docker 1.12.5 with calico v1.0.2 

### ISSUE THREAD

**Issue by sunilganatra** [projectcalico/calicoctl#1510]
I am trying to forward inbound traffic of host on following ports 22,10000.9443,8443 to calico network container. 

docker's port forwarding is not working as it is. I came across calico document to apply policy on calico-node. I tried to apply following policy but it is not working. could you please comment some examples on how to achieve port forwarding. 

- apiVersion: v1
  kind: policy
  metadata:
    name: port-forwarding
  spec:
    egress:
    - action: allow
    ingress:
    - action: allow
      destination:
        ports:
        - 22
        - 10000
        - 8443
        - 9443
      protocol: tcp
      source:

**Comment by tmjd** [comment]
Ok the first thing I feel I need to point out is that since you are using calico each container is receiving its own IP and doing port forwarding is not necessary since the containers can be reached directly.  If you are accessing the service from hosts that are not running calico node that those hosts will need to have the routes necessary to reach those IP addresses.

Second, I just asked the team I work with and port forwarding is not supported by calico.

With that said the following is a profile and policy that would work to open up the ports you have listed.  You will need a profile defined to set labels that the policy will use and then need to update your policy to use those labels. (Or you can just use profiles)  See http://docs.projectcalico.org/master/getting-started/docker/tutorials/advanced-policy

You need to load a profile similar to the following
```
cat << EOF | calicoctl apply -f -
- apiVersion: v1
  kind: profile
  metadata:
    name: <docker network name created with calico driver flags>
    labels:
      role: <policy label>
EOF
```

Then update your policy with something like
```
cat << EOF | calicoctl apply -f -
- apiVersion: v1
  kind: policy
  metadata:
    name: port-forwarding
  spec:
    selector: role == '<policy label>'
    egress:
    - action: allow
    ingress:
    - action: allow
      destination:
        ports:
        - 22
        - 10000
        - 8443
        - 9443
      protocol: tcp
      source: {}
EOF
```

Please update this issue if this works for you or you need further assistance.

**Comment by sunilganatra** [comment]
Hi Erik Stidham,

is there any option to achieve following usecase :

let's say calico docker network name is : dev and created with option --subnet 196.168.1.0/24 (not portable public ip subnet)

1. host A (running calico-node and containerA.  containerA  is associated to network **dev**. containerA has been created with docker port forwarding option [-p 9443:9443].  containerA is running httpd service on port 9443.)
2. host B (may or may not be running calico-node. may or may not have container associated to network **dev**.)

on host B, trying to wget file from containerA (wget http://hostA-IP:9443/file)

**Comment by tmjd** [comment]
To do that here is a great explanation from one of my colleagues (thanks @djosborne):
1. Get container routes onto the other host. You can do this by:
a. Configuring calico to peer with the networking fabric so that the other hosts' L3 hop already knows how to route to the containers. This is philiosphically the "best" solution but not always one people can implement if they're not running in private cloud and realistically not unless they're building production infrastructure.
b. Run a virtual bgp service on the host directly (such as BIRD). The cheating way to do this is just run calico/node on that host too, even if it's not going to run containers.
2. Set up a loadbalancer / proxy service to forward requests to the Calico IP. This is often orcehstrator dependent. For Mesos, use Marathon-LoadBalancer. For Kubernetes, there's service ports that can be used

**Comment by sunilganatra** [comment]
Thank you @tmjd and @djosborne. We will try bgp peering.
I have another question, if i want to reach container on some port(let's say 6767) through internet. how can i do that? If you have any example in any blog / doc please point me.

**Comment by tmjd** [comment]
This is the only document that talks about reaching a container externally but is light on content: 
http://docs.projectcalico.org/master/usage/external-connectivity#inbound-connectivity

I created an issue around possibly adding some documentation around what you are requesting since there is not much available right now.  Here is the issue https://github.com/projectcalico/calico/issues/509

I'm going to close this issue now because we have deviated significantly away from the original issue.  If you feel it should still be open please reopen or feel free to create a new issue if you have additional issues.

**Comment by sunilganatra** [comment]
@tmjd, thanks for providing links.  The link which you have given is pointing to BGP Peering (http://docs.projectcalico.org/master/usage/configuration/bgp) and i have already follow this. 

I tried example given in topic "Configuring a global BGP peer", is this the right topic to follow?
If it is than, it is not working for me ? could you please point me to right example.

**Comment by tmjd** [comment]
Yes that is the right topic to follow. I went through the exercise of setting up BIRD manually for a 2 node docker with calico cluster and was able to get it working, eventually (I'm not familiar with BIRD).  I have since torn down that setup or I would give you the configs I used.

As mentioned before an easy/cheaty method is to run calico/node on the host you are wanting to have routes to the containers in the cluster.  Also with this method you do not have to setup the 'global BGP peer' either.  To do that you would need docker on the host and then grab the `calicoctl` binary and then follow http://docs.projectcalico.org/v2.0/getting-started/docker/installation/vagrant-coreos/#install-calico.

One thing to remember, make sure when you are trying to access containers that you have configured permissions to allow the access you are attempting.  At one point I was trying to ping my containers but had only allowed tcp access on port 80.

One final thing I feel I should mention is that Tigera, the company I work for, is available to provide commercial support services. https://www.tigera.io/
